execution/function/function.go

```diff
@@ -1,247 +1,482 @@
+	series      []labels.Labels
+	once        sync.Once
+	vectorIndex int
+	nextOps     []model.VectorOperator
+
+	call FunctionCall
+
+	scalarInput  [][]model.StepVector
+	scalarPoints []float64
+}
+
+func NewfunctionOperator(funcExpr *parser.Call, call FunctionCall, nextOps []model.VectorOperator) (model.VectorOperator, error) {
+	f := &functionOperator{
+		nextOps:      nextOps,
+		call:         call,
+		funcExpr:     funcExpr,
+		vectorIndex:  0,
+		scalarInput:  make([][]model.StepVector, len(nextOps)-1),
+		scalarPoints: make([]float64, len(nextOps)-1),
+	}
+
+	for i := range funcExpr.Args {
+		if funcExpr.Args[i].Type() == parser.ValueTypeVector {
+			f.vectorIndex = i
+			break
+		}
+	}
+
+	// Check selector type.
+	// TODO(saswatamcode): Add support for string and matrix.
+	switch funcExpr.Args[f.vectorIndex].Type() {
+	case parser.ValueTypeVector, parser.ValueTypeScalar:
+		return f, nil
+	default:
+		return nil, errors.Wrapf(parse.ErrNotImplemented, "got %s:", funcExpr.String())
+	}
+}
+
+func (o *functionOperator) Explain() (me string, next []model.VectorOperator) {
+	return fmt.Sprintf("[*functionOperator] %v(%v)", o.funcExpr.Func.Name, o.funcExpr.Args), o.nextOps
+}
+
+func (o *functionOperator) Series(ctx context.Context) ([]labels.Labels, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	return o.series, nil
+}
+
+func (o *functionOperator) GetPool() *model.VectorPool {
+	return o.nextOps[o.vectorIndex].GetPool()
+}
+
+func (o *functionOperator) Next(ctx context.Context) ([]model.StepVector, error) {
+	if err := o.loadSeries(ctx); err != nil {
+		return nil, err
+	}
+
+	// Process non-variadic single/multi-arg instant vector and scalar input functions.
+	// Call next on vector input.
+	input, err := o.nextOps[o.vectorIndex].Next(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(input) == 0 {
+		return nil, nil
+	}
+
+	// Call next on all scalar inputs.
+	if len(o.scalarInput) != 0 {
+		s := 0
+		for i := range o.nextOps {
+			if i == o.vectorIndex {
+				continue
+			}
+
+			si, err := o.nextOps[i].Next(ctx)
+			if err != nil {
+				return nil, err
+			}
+
+			o.scalarInput[s] = si
+			defer o.nextOps[i].GetPool().PutVectors(o.scalarInput[s])
+			s++
+		}
+	}
+
+	for v := range input {
+		// scalar() depends on number of samples per vector and returns NaN if len(samples) != 1.
+		// So need to handle this separately here, instead of going via call which is per point.
+		if o.funcExpr.Func.Name == "scalar" {
+			if len(input[v].Samples) <= 1 {
+				continue
+			}
+
+			input[v].Samples = input[v].Samples[:1]
+			input[v].SampleIDs = input[v].SampleIDs[:1]
+			input[v].Samples[0] = math.Float64frombits(value.NormalNaN)
+			continue
+		}
+
+		for i := range input[v].Samples {
+			// Scalar expressions can select from storage as well.
+			if len(o.funcExpr.Func.ArgTypes) > 1 && len(o.scalarInput) > 0 {
+				for l := range o.scalarInput {
+					if len(o.scalarInput[l]) >= 1 {
```

Comment:
What happens if this condition is not met?

Reply:
I thought of adding it as a safeguard against empty input. But yup, a test case is necessary here I think. Might need to handle this differently. Empty series cases could also be a part of #88 I think. 🙂 
Will update this one

Reply:
I've added handling for this and verified empty cases manually. But a lot of it returns as NaN in certain cases. So will add test cases for this in #88